File: 3534-path-existence-queries-in-a-graph-ii/solution.py
class Solution:
    def answerQueries(self, n: int, nums: list[int], maxDiff: int, queries: list[list[int]]) -> list[int]:
        # Pair nodes with their values and sort by value
        sorted_nodes = sorted([(nums[i], i) for i in range(n)])
        
        comp_id = [0] * n
        current_comp = 0
        comp_id[sorted_nodes[0][1]] = current_comp
        
        # Identify connected components based on the maxDiff threshold
        for i in range(1, n):
            if sorted_nodes[i][0] - sorted_nodes[i-1][0] > maxDiff:
                current_comp += 1
            comp_id[sorted_nodes[i][1]] = current_comp
            
        results = []
        for u, v in queries:
            if u == v:
                results.append(0)
            elif comp_id[u] != comp_id[v]:
                results.append(-1)
            else:
                # If in same component, distance is 1 if directly reachable, else 2
                if abs(nums[u] - nums[v]) <= maxDiff:
                    results.append(1)
                else:
                    results.append(2)
        return results

File: 3534-path-existence-queries-in-a-graph-ii/solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

class Solution {
public:
    std::vector<int> answerQueries(int n, std::vector<int>& nums, int maxDiff, std::vector<std::vector<int>>& queries) {
        struct Node { int val, id; };
        std::vector<Node> sortedNodes(n);
        for (int i = 0; i < n; ++i) sortedNodes[i] = {nums[i], i};
        
        std::sort(sortedNodes.begin(), sortedNodes.end(), [](const Node& a, const Node& b) {
            return a.val < b.val;
        });

        std::vector<int> compId(n);
        int currentComp = 0;
        compId[sortedNodes[0].id] = currentComp;

        for (int i = 1; i < n; ++i) {
            if (sortedNodes[i].val - sortedNodes[i - 1].val > maxDiff) {
                currentComp++;
            }
            compId[sortedNodes[i].id] = currentComp;
        }

        std::vector<int> results;
        results.reserve(queries.size());
        for (const auto& q : queries) {
            int u = q[0], v = q[1];
            if (u == v) {
                results.push_back(0);
            } else if (compId[u] != compId[v]) {
                results.push_back(-1);
            } else {
                results.push_back(std::abs(nums[u] - nums[v]) <= maxDiff ? 1 : 2);
            }
        }
        return results;
    }
};